src/Orleans.Transactions/DeadlockDetection/DeadlockDetector.cs

```diff
@@ -76,47 +80,88 @@ public Batch(IEnumerable<SiloAddress> silos)
+            this.options = options.Value;
         }
 
         public override async Task OnActivateAsync()
         {
             await base.OnActivateAsync();
             this.siloStatusOracle = this.ServiceProvider.GetRequiredService<ISiloStatusOracle>();
             this.internalGrainFactory = this.ServiceProvider.GetRequiredService<IInternalGrainFactory>();
+            this.deadlockListeners = this.ServiceProvider.GetServices<IDeadlockListener>().ToArray();
         }
 
         public async Task CheckForDeadlocks(CollectLocksResponse message)
         {
-            this.logger.LogInformation($"CheckForDeadlocks({message.BatchId},{message.SiloAddress},{message.MaxVersion})");
+            // TODO LOG  this.logger.LogInformation($"CheckForDeadlocks({message.BatchId},{message.SiloAddress},{message.MaxVersion})");
             Batch batch;
             if (message.BatchId == null)
             {
-                // new batch!
-                batch = new Batch(this.GetSiloAddresses());
-                this.batches[batch.Id] = batch;
+                if (this.StartNewBatch(message, out batch))
+                {
+                    this.batches[batch.Id] = batch;
+                }
+                else
+                {
+                    // TODO log rate limited
+                    return;
+                }
             }
             else if(!this.batches.TryGetValue(message.BatchId.Value, out batch))
             {
-                if(this.logger.IsEnabled(LogLevel.Information)){ this.logger.LogInformation($"received message for missing batch {message}"); }
-
-                return;
+               // TODO LOG  if(this.logger.IsEnabled(LogLevel.Trace)){ this.logger.LogInformation($"received message for missing batch {message}"); }
+               return;
             }
 
             await this.UpdateBatch(batch, message);
         }
 
-        private async Task UpdateBatch(Batch batch, CollectLocksResponse message)
+        private bool StartNewBatch(CollectLocksResponse message, out Batch batch)
+        {
+            if (this.batches.Count >= this.options.MaxConcurrentDeadlockAnalysis)
+            {
+                // TODO report?
+                this.logger.LogWarning("Too many batches exist, discarding a message");
+                batch = null;
+                return false;
+            }
+
+            // look for a batch that intersects the grains / transactions in this message
+            // before starting a new one
+            var transactionsInMessage = message.Locks.Select(l => l.TxId).ToSet();
+            foreach (var existingBatch in this.batches.Values)
+            {
+                this.logger.LogInformation(
+                    $"checking for overlap between messageTxs {string.Join(",", transactionsInMessage)}" +
+                    $" and batch tx {string.Join(",", existingBatch.KnownTransactions)}");
+                if (existingBatch.KnownTransactions.Overlaps(transactionsInMessage))
```

Comment:
This is only detecting overlaps in transactions not in resources look the comment above says